server/storage/bootstrap.go

```diff
@@ -0,0 +1,335 @@
+						zap.Uint64("snapshot-index", snapshot.Metadata.Index),
+					)
+				}
+			}
+		} else {
+			cfg.Logger.Info("No snapshot found. Recovering WAL from scratch!")
+		}
+	}
+	return &bootstrappedBackend{
+		BeHooks:  beHooks,
+		Be:       be,
+		Ci:       ci,
+		BeExist:  beExist,
+		Snapshot: snapshot,
+	}, nil
+}
+
+func maybeDefragBackend(cfg config.ServerConfig, be backend.Backend) error {
+	size := be.Size()
+	sizeInUse := be.SizeInUse()
+	freeableMemory := uint(size - sizeInUse)
+	thresholdBytes := cfg.ExperimentalBootstrapDefragThresholdMegabytes * 1024 * 1024
+	if freeableMemory < thresholdBytes {
+		cfg.Logger.Info("Skipping defragmentation",
+			zap.Int64("current-db-size-bytes", size),
+			zap.String("current-db-size", humanize.Bytes(uint64(size))),
+			zap.Int64("current-db-size-in-use-bytes", sizeInUse),
+			zap.String("current-db-size-in-use", humanize.Bytes(uint64(sizeInUse))),
+			zap.Uint("experimental-bootstrap-defrag-threshold-bytes", thresholdBytes),
+			zap.String("experimental-bootstrap-defrag-threshold", humanize.Bytes(uint64(thresholdBytes))),
+		)
+		return nil
+	}
+	return be.Defrag()
+}
+
+// BootstrapWALFromSnapshot reads the WAL at the given snap and returns the Wal, its latest HardState and cluster ID, and all entries that appear
+// after the position of the given snap in the WAL.
+// The snap must have been previously saved to the WAL, or this call will panic.
+func BootstrapWALFromSnapshot(cfg config.ServerConfig, snapshot *raftpb.Snapshot) *BootstrappedWAL {
+	var walsnap walpb.Snapshot
+	if snapshot != nil {
+		walsnap.Index, walsnap.Term = snapshot.Metadata.Index, snapshot.Metadata.Term
+	}
+	repaired := false
+	for {
+		w, err := wal.Open(cfg.Logger, cfg.WALDir(), walsnap)
+		if err != nil {
+			cfg.Logger.Fatal("failed to open WAL", zap.Error(err))
+		}
+		if cfg.UnsafeNoFsync {
+			w.SetUnsafeNoFsync()
+		}
+		wmetadata, st, ents, err := w.ReadAll()
+		if err != nil {
+			w.Close()
+			// we can only repair ErrUnexpectedEOF and we never repair twice.
+			if repaired || err != io.ErrUnexpectedEOF {
+				cfg.Logger.Fatal("failed to read WAL, cannot Be repaired", zap.Error(err))
+			}
+			if !wal.Repair(cfg.Logger, cfg.WALDir()) {
+				cfg.Logger.Fatal("failed to repair WAL", zap.Error(err))
+			} else {
+				cfg.Logger.Info("repaired WAL", zap.Error(err))
+				repaired = true
+			}
+			continue
+		}
+		var m etcdserverpb.Metadata
+		pbutil.MustUnmarshal(&m, wmetadata)
+		id := types.ID(m.NodeID)
+		cid := types.ID(m.ClusterID)
+		meta := &Metadata{ClusterID: cid, NodeID: id}
+
+		wal := &BootstrappedWAL{
+			lg:       cfg.Logger,
+			HaveWAL:  true,
+			Wal:      w,
+			State:    &st,
+			Entries:  ents,
+			Snapshot: snapshot,
+			Meta:     meta,
+		}
+
+		if cfg.ForceNewCluster {
+			// discard the previously uncommitted entries
+			wal.Entries = wal.CommitedEntries()
+			entries := wal.ConfigChangeEntries(meta)
+			// force commit config change entries
+			wal.AppendAndCommitEntries(entries)
```

Comment:
If we are creating a new clusters (it's how I understand 'ForceNew') I would expect using newly generated ClusterId & NodeId, while this (and pre-refactoring logic) seems to reuse ID from the source Snapshot... a little confused... 
just flagging.

Reply:
Please check out old logic https://github.com/etcd-io/etcd/blob/946a5a6f25c3b6b89408ab447852731bde6e6289/server/etcdserver/raft.go#L516-L586

Fact that we are forcing new cluster has no impact on WAL bootstrap. It's the call to RestartRaft that happens after does the generation of new ClusterID and NodeID. https://github.com/etcd-io/etcd/blob/946a5a6f25c3b6b89408ab447852731bde6e6289/server/etcdserver/raft.go#L583